04-onion-routing.md

```diff
@@ -367,6 +379,112 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Route Blinding
+
+Nodes receiving onion packets may hide their identity from senders by
+"blinding" an arbitrary amount of hops at the end of an onion path.
+
+When using route blinding, nodes find a route to themselves from a given
+"introduction node". They then use ECDH with each node in that route to create
+a "blinded" node ID and an encrypted blob (`encrypted_recipient_data`) for each
+one of the blinded nodes.
+
+They communicate this blinded route and the encrypted blobs to the sender.
+The sender finds a route to the introduction node and extends it with the
+blinded route provided by the recipient. The sender includes the encrypted
+blobs in the corresponding onion payloads: they allow nodes in the blinded
+part of the route to "unblind" the next node and correctly forward the packet.
+
+The `encrypted_recipient_data` is a TLV stream, encrypted for a given blinded
+node, that may contain the following TLV fields:
+
+1. `tlv_stream`: `encrypted_recipient_data_tlv`
+2. types:
+    1. type: 2 (`short_channel_id`)
+    2. data:
+        * [`short_channel_id`:`short_channel_id`]
+    1. type: 4 (`next_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+
+#### Requirements
+
+A recipient N(r) creating a blinded route `N(0) -> N(1) -> ... -> N(r)` to itself:
+
+- MUST create a blinded node ID `B(i)` for each node using the following algorithm:
+  - `e(0) <- {0;1}^256`
+  - `E(0) = e(0) * G`
+  - For every node `N(i)` in the route:
+    - let `P(i) = k(i) * G` be `N(i)`'s `node_id` (`k(i)` is `N(i)`'s private key)
+    - `ss(i) = H(e(i) * P(i)) = H(k(i) * E(i))` (shared secret known only by `N(r)` and `N(i)`)
+    - `B(i) = HMAC256("blinded_node_id", ss(i)) * P(i)` (blinded `node_id` for `N(i)`, private key known only by `N(i)`)
+    - `rho(i) = HMAC256("rho", ss(i))`
+    - `e(i+1) = H(E(i) || ss(i)) * e(i)` (ephemeral private key, only known by `N(r)`)
+    - `E(i+1) = H(E(i) || ss(i)) * E(i)` (NB: `N(i)` MUST NOT learn `e(i)`)
+- MAY create an `encrypted_recipient_data` for each blinded node `B(i)`
+- If it creates `encrypted_recipient_data`:
+  - MUST encrypt it with ChaCha20-Poly1305 using the `rho(i)` key and an all-zero nonce
+  - MUST use `E(i)` in ChaCha20-Poly1305's associated authenticated data
```

Comment:
This requirement is weird?   E(i) is already a requirement since it's the key, adding it to AD here seem voodoo?  (I assumed no AD was required here).

Reply:
Right, for nodes after the introduction point, if `E(i)` was invalid the onion decryption would fail, so we would detect an invalid `E(i)` before even trying to decrypt the `encrypted_recipient_data`, so this check is unnecessary.

I believed it was necessary for the introduction point (because its onion is *not* blinded), to ensure that in case an invalid `E(i)` was provided an error would come from the introduction point and not the next node. But that also seems unnecessary, because with an invalid `E(i)` the introduction point wouldn't be able to decrypt its `encrypted_recipient_data` anyway. I'll dive into it again and will remove if I don't find a good reason to keep it.

Reply:
Done in https://github.com/lightningnetwork/lightning-rfc/pull/765/commits/cc855c754bca1ac08e8efc9f27a3c4b86773d953